autogpts/autogpt/autogpt/agents/base.py

```diff
@@ -199,198 +245,96 @@ def llm(self) -> ChatModelInfo:
-            ),
-            extra_messages=extra_messages,
-            **extras,
-        )
-
-        return prompt
-
-    def on_before_think(
-        self,
-        prompt: ChatPrompt,
-        scratchpad: PromptScratchpad,
-    ) -> ChatPrompt:
-        """Called after constructing the prompt but before executing it.
-
-        Calls the `on_planning` hook of any enabled and capable plugins, adding their
-        output to the prompt.
-
-        Params:
-            prompt: The prompt that is about to be executed.
-            scratchpad: An object for plugins to write additional prompt elements to.
-                (E.g. commands, constraints, best practices)
-
-        Returns:
-            The prompt to execute
-        """
-        current_tokens_used = self.llm_provider.count_message_tokens(
-            prompt.messages, self.llm.name
-        )
-        plugin_count = len(self.config.plugins)
-        for i, plugin in enumerate(self.config.plugins):
-            if not plugin.can_handle_on_planning():
-                continue
-            plugin_response = plugin.on_planning(scratchpad, prompt.raw())
-            if not plugin_response or plugin_response == "":
-                continue
-            message_to_add = ChatMessage.system(plugin_response)
-            tokens_to_add = self.llm_provider.count_message_tokens(
-                message_to_add, self.llm.name
-            )
-            if current_tokens_used + tokens_to_add > self.send_token_limit:
-                logger.debug(f"Plugin response too long, skipping: {plugin_response}")
-                logger.debug(f"Plugins remaining at stop: {plugin_count - i}")
+    # Generic function to get components of a specific type
+    def get_component(self, component_type: type[T]) -> Optional[T]:
+        for component in self.components:
+            if isinstance(component, component_type):
+                return component
+        return None
+
+    def _selective_copy(self, args: tuple[Any, ...]) -> tuple[Any, ...]:
+        copied_args = []
+        for item in args:
+            if isinstance(item, list):
+                # Shallow copy for lists
+                copied_item = item[:]
+            elif isinstance(item, dict):
+                # Shallow copy for dicts
+                copied_item = item.copy()
+            elif isinstance(item, BaseModel):
+                # Deep copy for Pydantic models (deep=True to also copy nested models)
+                copied_item = item.copy(deep=True)
+            else:
+                # Deep copy for other objects
+                copied_item = copy.deepcopy(item)
+            copied_args.append(copied_item)
+        return tuple(copied_args)
+
+    def is_enabled(self, component: Component) -> bool:
+        if callable(component.enabled):
+            return component.enabled()
+        return component.enabled
+
+    async def foreach_components(
+        self, method_name: str, *args, retry_limit: int = 3
+    ) -> Any:
```

Comment:
This is the core of component execution. This visits components that implement the method and executes the code, handles exceptions and retires. Looking for feedback how could I make this more type safe.